### DISCUSSION TITLE
XML PATH

### DISCUSSION THREAD

**Discussion by gdiana** [grafana/grafana-infinity-datasource#27] [Q&A] [answered]
**Describe the bug**
Hi, I am experiencing a problem using XML. I am loading data from a URL and for whatever reason I am unable to filter the XML data. 

Here is a sample of the data. Normally it returns many samples depending on the interval of the URL query. Sorry had to send as image due to formatting problems.

![image](https://user-images.githubusercontent.com/13445152/104828291-680a0280-5870-11eb-944d-9f0c1fa98819.png)

**Screenshots**
Below is a screenshot and as may be seen no data is being returned despite the correct Rows/Root Path

![image](https://user-images.githubusercontent.com/13445152/104828123-92f35700-586e-11eb-97d4-df9ca0d0deed.png)

If I repeat this inline I get same but if I strip off the xml and meter outer layers then it works fine as shown below

![image](https://user-images.githubusercontent.com/13445152/104828315-c7681280-5870-11eb-8d75-997696553db0.png)

![image](https://user-images.githubusercontent.com/13445152/104828242-c1bdfd00-586f-11eb-984c-f5b5a714baca.png)

**Version Details: 
 - Grafana version : 7.3.6  (ea06633c34)
 - Plugin version :   0.6.0

**Additional context**

I have not used it for anything else but it appears the number of levels might be an issue. As I am getting data from a 3 party I cannot change the xml output.

But hey this is a great tool and it might be useful to consider a JASONATA go between so that one may query and reformat any URL return data and massage the output to get into the correct format for GRAFANA somewhat like NODE-RED does.

If I convert the above XML data to JSON it works fine using the same Rows/Root (xml.meter.profile.sample) as shown below. Unfortunately the provider only returns XML data and not JSON?

![image](https://user-images.githubusercontent.com/13445152/104828513-e7003a80-5872-11eb-9bea-b69e61266db6.png)

This is the XML Tree for the stripped down data that works
![image](https://user-images.githubusercontent.com/13445152/104828544-3f373c80-5873-11eb-8c2b-22f73fc82f8e.png)

and this is the XML tree for the full data set. As you can see there is a slight difference in that the full set is an array with index 0----57517
![image](https://user-images.githubusercontent.com/13445152/104828563-6e4dae00-5873-11eb-8bb9-8077f1f07292.png)

Please shout if you need any additional information.

Kind regards

Greg Diana

**Comment by ghost** [comment]
same here:

```
<root>
    <text>level 1</text>
    <level2>
        <text>level 2</text>
        <level3>
            <text>level 3</text>
        </level3>
    </level2>
</root>
```
text shows for root, root.level1, and root.level1.level2 but not when adding level3.
and it still does not work when removing the root and using 'level1.level2.level3' as root.

**Comment by yesoreyeram** [comment]
@gdiana @user-cw  - sorry for the delay. See below example helps. Actually the issue is with the root element selection. By default xml will consider level of node to be array. So you need to specify index if you have parent node.

### Case 1 :

```xml
<profile>
    <sample>
        <date>2021-02-04 19:23:08</date>
        <p1>1000</p1>
        <p2>2000</p2>
    </sample>
    <sample>
        <date>2021-02-04 19:24:08</date>
        <p1>1020</p1>
        <p2>1050</p2>
    </sample>
    <sample>
        <date>2021-02-04 19:25:08</date>
        <p1>1010</p1>
        <p2>2030</p2>
    </sample>
</profile>
```
![image](https://user-images.githubusercontent.com/153843/106944936-1e9a2e00-671f-11eb-91b7-be096e924b55.png)


### Case 2

```xml
<xml>
<profile>
    <sample>
        <date>2021-02-04 19:23:08</date>
        <p1>1000</p1>
        <p2>2000</p2>
    </sample>
    <sample>
        <date>2021-02-04 19:24:08</date>
        <p1>1020</p1>
        <p2>1050</p2>
    </sample>
    <sample>
        <date>2021-02-04 19:25:08</date>
        <p1>1010</p1>
        <p2>2030</p2>
    </sample>
</profile>
</xml>
```
![image](https://user-images.githubusercontent.com/153843/106945128-5acd8e80-671f-11eb-9fb1-876ce940ff2d.png)

I am closing this issue. Feel free report it / reopen if the issue exists.

**Comment by gdiana** [comment, marked as answer]
Thank you, much appreciated.

Noting that the xml data was formatted as follows

![image](https://user-images.githubusercontent.com/13445152/106956431-40a3a880-673f-11eb-84d1-297022a6e675.png)

I used  **xml.meter[0].profile[0].sample** and it worked great as shown below :)

![image](https://user-images.githubusercontent.com/13445152/106956684-a132e580-673f-11eb-94a3-c977475caa30.png)

Just for interest when I first tried this it gave some weird results but after a few times the Query inspector started to return the correct data and it no works fine. Seems like stuff is cached somewhere. Weird but true.

**Comment by yesoreyeram** [comment]
Glad to see it worked for you @gdiana . Yea. Might be due to caching. Thanks for testing the plugin and providing the feedback. Any feedback/issues/feature requests are much appreciated.

**Reply by gdiana** [reply]
No problem and keep up the good work.
